java/ql/src/experimental/Security/CWE/CWE-078/CommandInjectionRuntimeExec.qll

```diff
@@ -0,0 +1,170 @@
+
+  override predicate isSink(DataFlow::Node sink) {
+    exists(RuntimeExecMethod method, MethodAccess call |
+      call.getMethod() = method and
+      sink.asExpr() = call.getArgument(_) and
+      sink.asExpr().getType() instanceof Array
+    )
+  }
+
+  override predicate isSanitizer(DataFlow::Node node) {
+    node.asExpr().getFile().isSourceFile() and
+    (
+      node.getType() instanceof PrimitiveType or
+      node.getType() instanceof BoxedType
+    )
+  }
+}
+
+module ExecUserFlowConfig implements DataFlow::ConfigSig {
+  predicate isSource(DataFlow::Node source) {
+    source instanceof Source
+  }
+
+  predicate isSink(DataFlow::Node sink) {
+    exists(RuntimeExecMethod method, MethodAccess call |
+      call.getMethod() = method and
+      sink.asExpr() = call.getArgument(_) and
+      sink.asExpr().getType() instanceof Array
+    )
+  }
+
+  predicate isBarrier(DataFlow::Node node) {
+    node.asExpr().getFile().isSourceFile() and
+    (
+      node.getType() instanceof PrimitiveType or
+      node.getType() instanceof BoxedType
+    )
+  }
+}
+
+/** Tracks flow of unvalidated user input that is used in Runtime.Exec */
+module ExecUserFlow = TaintTracking::Global<ExecUserFlowConfig>;
+
+// array[3] = node
+class AssignToNonZeroIndex extends DataFlow::Node {
+  AssignToNonZeroIndex() {
+    exists(AssignExpr assign, ArrayAccess access |
+      assign.getDest() = access and
+      access.getIndexExpr().(IntegerLiteral).getValue() != "0" and
+      assign.getSource() = this.asExpr()
+    )
+  }
+}
+
+// String[] array = {"a", "b, "c"};
+class ArrayInitAtNonZeroIndex extends DataFlow::Node {
+  ArrayInitAtNonZeroIndex() {
+    exists(ArrayInit init, int index |
+      init.getInit(index) = this.asExpr() and
+      index != 0
+    )
+  }
+}
+
+// Stream.concat(Arrays.stream(array_1), Arrays.stream(array_2))
+class StreamConcatAtNonZeroIndex extends DataFlow::Node {
+  StreamConcatAtNonZeroIndex() {
+    exists(MethodAccess call, int index |
+      call.getMethod().getQualifiedName() = "java.util.stream.Stream.concat" and
+      call.getArgument(index) = this.asExpr() and
+      index != 0
+    )
+  }
+}
+
+// allow list of executables that execute their arguments
+// TODO: extend with data extensions
+class UnSafeExecutable extends string {
+  bindingset[this]
+  UnSafeExecutable() {
+    this.regexpMatch("^(|.*/)([a-z]*sh|javac?|python.*|perl|[Pp]ower[Ss]hell|php|node|deno|bun|ruby|osascript|cmd|Rscript|groovy)(\\.exe)?$") and
+    not this = "netsh.exe"
+  }
+}
+
+predicate callIsTaintedByUserInputAndDangerousCommand(MethodAccess call, ExecUserFlow::PathNode source, ExecUserFlow::PathNode sink, DataFlow::Node sourceCmd, DataFlow::Node sinkCmd) {
+  call.getMethod() instanceof RuntimeExecMethod and
+  // this is a command-accepting call to exec, e.g. rt.exec(new String[]{"/bin/sh", ...})
+  (
+    ExecCmdFlow::flow(sourceCmd, sinkCmd) and
+    sinkCmd.asExpr() = call.getArgument(0)
```

Comment:
Be careful: a varargs argument isn't a `DataFlow::Node`, but rather a `DataFlow::ImplicitVarargsArray`, so `Node::asExpr` won't hold in certain cases. Refer to `ExternalProcess::argumentToExec` for a more complete implementation:

https://github.com/github/codeql/blob/0a0e34716e6e1c228284ffa24a598ea915b1da63/java/ql/lib/semmle/code/java/security/ExternalProcess.qll#L31-L36

Reply:
I don't _think_ that `java.lang.Runtime.exec()` can take varargs.

I just checked the [function signature](https://docs.oracle.com/en/java/javase/18/docs/api/java.base/java/lang/Runtime.html#exec(java.lang.String%5B%5D)) and it looks like `exec(String[] cmdarray)`, etc. without any `...`.

It would be hard for it to take varargs, since it can also take `envp` and `dir` arguments after the command array, and Java can only accept varargs in the final argument.

Reply:
Hm, you're right, I was thinking in more general terms without realizing this applies to a very specific sink. Feel free to ignore my comment then — we can always apply this if the query is expanded to cover other sinks.